### PR TITLE
State and prove the universal property of the free magma quotiented by a theory.

### LINKED ISSUE #412: CONFLUENT: Formalize and prove Theorem 9.2 (existence and uniqueness of free magmas subject to a theory)
This is a generalization of #133 and so one may be able to adapt the code.

Of course, an initial step will be to formalize what it means to be a free magma relative to a theory in the first place (Definition 9.1).

### PR DESCRIPTION
**Pull request by codyroux**
null

Closes #412

### CODE REVIEW

File: equational_theories/Completeness.lean [hunk 1]
```diff
@@ -156,3 +156,71 @@ theorem Completeness' {α β} {Γ : Ctx α} {E : MagmaLaw β} (h : Γ ⊧ E) : N
 theorem Completeness {α} {Γ : Ctx α} {E : MagmaLaw α} (h : Γ ⊧ E) : Nonempty (Γ ⊢ E) :=
   match Completeness' h with
   | .intro x => .intro (derive_of_derive' x)
+
+#check ⟦3⟧
+
+#print Soundness'
+#print satisfiesPhi
```
Comment (pitmonticone) [inline equational_theories/Completeness.lean:163]: Can you remove these `#check` and `#print`.

File: equational_theories/Completeness.lean [hunk 2]
```diff
@@ -156,3 +156,71 @@ theorem Completeness' {α β} {Γ : Ctx α} {E : MagmaLaw β} (h : Γ ⊧ E) : N
...
+
+def FreeMagmaWithLaws.eval {α G} {Γ : Ctx α} (φ : α → G) [Magma G] (modelsG : G ⊧ Γ) :
+FreeMagmaWithLaws α Γ → G :=
+Quotient.lift (evalInMagma φ) (
+by
+  intros a b
+  simp [HasEquiv.Equiv, SetoidOfLaws, RelOfLaws]
```
Comment (pitmonticone) [inline equational_theories/Completeness.lean:170]: This non-terminal `simp` should be avoided. Please use the `simp only` output of `simp?`.

File: equational_theories/Completeness.lean [hunk 3]
```diff
@@ -156,3 +156,71 @@ theorem Completeness' {α β} {Γ : Ctx α} {E : MagmaLaw β} (h : Γ ⊧ E) : N
...
+  -- hmpf choice again.
+  have ⟨ x_bar, eqx ⟩ := Quotient.exists_rep x
+  have ⟨ y_bar, eqy ⟩ := Quotient.exists_rep y
+  rw [← eqx, ← eqy, Quotient.lift₂_mk]
+  repeat rw [Quotient.lift_mk]
+  simp [evalInMagma]
+
+lemma eq_app : ∀ α β (f g : α → β), f = g → ∀ x, f x = g x :=
+by
+  exact fun α β f g a x ↦ congrFun a x
```
Comment (pitmonticone) [inline equational_theories/Completeness.lean:193]: Please commit this suggestion:

```suggestion
lemma eq_app : ∀ α β (f g : α → β), f = g → ∀ x, f x = g x := fun α β f g a x ↦ congrFun a x
```

**Review by pitmonticone — CHANGES_REQUESTED** [review]

**Comment by codyroux** [comment]
Thanks Pietro!

**Comment by pitmonticone** [comment]
Thank you @codyroux for your contribution!

**Pull request merged**